basetest.pm

```diff
@@ -449,17 +455,19 @@ sub record_serialresult {
     $output .= "# Result:\n";
     $output .= "$string\n";
     $self->record_resultfile('wait_serial', $output, result => $res);
-    return;
+    return undef;
```

Comment:
Just wondering, why is this needed?

Reply:
in short: because @kraih said so :)

Reply:
I assume the intention behind the `return;` was to always return `undef` (regardless of the last expression). However, using only `return;` would actually return an empty list if the function is executed where an array is wanted. We likely don't want that behavior.

Reply:
That's why I asked. I always preferred to call plain `return;` as then you get empty list, where with `return undef;` you get a list with one `undef` element IIUC. And if I greped source correctly I found that `record_serialresult` and `record_soft_failure_result` are never called in tests nor is their return actually used. I see only `register_extra_test_results` is used somewhere in other returns.
So I was wondering if there is something new I missed.

Reply:
This change is correct, bare `return;` is not considered good Perl anymore. In recent years we've seen too many high profile security issues with functions/methods that behave differently depending on the context in which they were used. This is usually caused by `wantarray`, but bare `return;` has similar risks. So just don't make functions/methods that mix contexts, return a list or a scalar, never both. (The Perl Best Practices book does recommend bare `return;`, but the author now also considers this an error.)

Reply:
@aaannz Yes, the return value isn't actually used here. But I'd like to introduce the currently preferred best practices so hopefully they are copied/used more often.

Reply:
The [Modern Perl Best Practices](https://learning.oreilly.com/videos/modern-perl-best/9781491910719) training also mentions this.

Reply:
@kraih Perfect. This is what I didn't know and I still lived in the old world where `return;` was recommended. Thanks.